exporters/otlp/otlplog/otlploggrpc/internal/observ/instrumentation.go

```diff
@@ -0,0 +1,269 @@
+// RejectedItems field of the PartialSuccess.
+func (e ExportOp) End(err error) {
+	addOpt := get[metric.AddOption](addOpPool)
+	defer put(addOpPool, addOpt)
+
+	*addOpt = append(*addOpt, e.inst.setOpt)
+
+	success := successful(e.nLogs, err)
+
+	e.inst.logInflightMetric.Add(e.ctx, -e.nLogs, *addOpt...)
+	e.inst.logExportedMetric.Add(e.ctx, success, *addOpt...)
+
+	mOpt := e.inst.setOpt
+	if err != nil {
+		// Add the error.type attribute to the attribute set.
+		attrs := get[attribute.KeyValue](attrsPool)
+		defer put(attrsPool, attrs)
+		*attrs = append(*attrs, e.inst.presetAttrs...)
+		*attrs = append(*attrs, semconv.ErrorType(err))
+
+		set := attribute.NewSet(*attrs...)
+		mOpt = metric.WithAttributeSet(set)
+
+		// Reset addOpt with new attribute set
+		*addOpt = append((*addOpt)[:0], mOpt)
+
+		e.inst.logExportedMetric.Add(e.ctx, e.nLogs-success, *addOpt...)
+	}
+
+	code := status.Code(err)
+
+	recordOpt := get[metric.RecordOption](recordOptPool)
+	defer put(recordOptPool, recordOpt)
+	*recordOpt = append(
+		*recordOpt,
+		mOpt,
+		metric.WithAttributes(
+			semconv.RPCGRPCStatusCodeKey.Int64(int64(code)),
+		),
+	)
+	duration := time.Since(e.start).Seconds()
+	e.inst.logExportedDurationMetric.Record(e.ctx, duration, *recordOpt...)
+}
+
+// successful returns the number of successfully exported logs out of the n
+// that were exported based on the provided error.
+//
+// If err is nil, n is returned. All logs were successfully exported.
+//
+// If err is not nil and not an [internal.PartialSuccess] error, 0 is returned.
+// It is assumed all logs failed to be exported.
+//
+// If err is an [internal.PartialSuccess] error, the number of successfully
+// exported logs is computed by subtracting the RejectedItems field from n. If
+// RejectedItems is negative, n is returned. If RejectedItems is greater than
+// n, 0 is returned.
+func successful(n int64, err error) int64 {
+	if err == nil {
+		return n // All logs successfully exported.
+	}
+	// Split rejection calculation so successful is inlineable.
+	return n - rejectedCount(n, err)
+}
+
+var errPool = sync.Pool{
+	New: func() any {
+		return new(internal.PartialSuccess)
+	},
+}
+
+// rejectedCount returns how many out of the n logs exporter were rejected based on
+// the provided non-nil err.
+func rejectedCount(n int64, err error) int64 {
+	ps := errPool.Get().(*internal.PartialSuccess)
+	defer errPool.Put(ps)
+
+	// check for partial success
+	if errors.As(err, ps) {
+		return min(max(ps.RejectedItems, 0), n)
+	}
+	// all logs exporter
+	return n
+}
+
+// ServerAddrAttrs is a function that extracts server address and port attributes
+// from a target string.
+func ServerAddrAttrs(target string) []attribute.KeyValue {
+	addr, port, err := ParseCanonicalTarget(target)
+	if err != nil {
+		return []attribute.KeyValue{semconv.ServerAddress(target)}
+	}
+	if port == -1 {
+		return []attribute.KeyValue{semconv.ServerAddress(addr)}
+	}
```

Comment:
It is not guaranteed that the host is not empty.

```suggestion
	if port == -1 && addr != "" {
		return []attribute.KeyValue{semconv.ServerAddress(addr)}
	}
```